File: Part 2 - Regression/Section 6 - Polynomial Regression/Polynomial_Regression_Implementation.R
# Data Preprocessing Template

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# # Splitting the dataset into the Training set and Test set
# # install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling

# !We won't need feature scaling since a polynomial regression model is a 
#  a multiple linear regression model with polynomial terms. !!!!!

####-----------------------------------------------------------------------###
####                  Fitting two models to our data                       ###
####-----------------------------------------------------------------------###


#Fitting Linear Regression Model.
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

#Fitting Polynomial Regression Model.
#We create an additional set of features that we will input to our multiple linear
#regression model. 
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4

poly_reg = lm(formula = Salary ~ .,
              data = dataset) 

summary(poly_reg)


####-----------------------------------------------------------------------###
####                  Visualizing the Models - Linear Regression Model     ###
####-----------------------------------------------------------------------###

ggplot() + geom_point(aes(x=dataset$Level , y=dataset$Salary),
                      colour = 'red') +
geom_line(aes(x=dataset$Level , y=predict(lin_reg, new_data=dataset)),
          colour = 'blue') + 
          ggtitle('Truth or Bluff(Linear Regression)') +
          xlab('Levels') +
          ylab('Salary')

####-----------------------------------------------------------------------###
####                  Visualizing the Models - Polynomial Model            ###
####-----------------------------------------------------------------------###

ggplot() + geom_point(aes(x=dataset$Level , y=dataset$Salary),
                      colour = 'red') +
  geom_line(aes(x=dataset$Level, y=predict(poly_reg, new_data=dataset)),
            colour = 'blue') + 
  ggtitle('Truth or Bluff(Polynomial Model)') +
  xlab('Levels') +
  ylab('Salary')


####-----------------------------------------------------------------------###
####                  Predicting a new result - Linear Regression          ###
####-----------------------------------------------------------------------###
y_pred = predict(lin_reg, data.frame(Level=6.5))

####-----------------------------------------------------------------------###
####                  Predicting a new result - Polynomial Model           ###
####-----------------------------------------------------------------------###

y_pred = predict(poly_reg, data.frame(Level=6.5),
                                      Level2=6.5^2,
                                      Level3=6.5^3,
                                      Level4=6.5^4))


                                      

File: Part 02 - Regression/Section 6 - Polynomial Regression/Polynomial_Regression_Implementation.py

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')

X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

 
#We dont need feature scaling because polynomial regression consists of
#adding some polynomial terms into the multiple linear regression
#equation and therefore we will use the same l.regression library as
#the one we used when we built our linear and multiple regression 
#models. 

#Fitting linear regression model to the dataset

from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,y)

from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree=4)
X_poly = poly_reg.fit_transform(X)

#Plotting the linear regression results
lin_reg2 = LinearRegression()
lin_reg2.fit(X_poly,y)
plt.scatter(X,y, color = 'red')
plt.plot(X, lin_reg.predict(X), color = 'blue')
plt.title("Truth or Bluff(Linear Regression)")
plt.xlabel('Position Label')
plt.ylabel('Salary')
plt.show()

#Plotting the polynomial regression
#We use fit_transform method to fit the data X to our polynomial function.
#-----------------------------------------------------#
##We create a new X matrix to create a better plot, with a higher resolution
##Instead of step=1 now we want step = 0.1

X_grid = np.arange(min(X),max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid),1))
#-----------------------------------------------------#
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, lin_reg2.predict(poly_reg.fit_transform(X_grid)), color = 'orange')
plt.title('Polynomial Regression Model')
plt.xlabel('Position Label')
plt.ylabel('Salary')
plt.show()

########
lin_reg.predict(6.5)
lin_reg2.predict(poly_reg.fit_transform(6.5)

########






